src/library_dylink.js

```diff
@@ -1153,26 +1153,34 @@ var LibraryDylink = {
   // Async version of dlopen.
   _emscripten_dlopen_js__deps: ['$dlopenInternal', '$callUserCallback', '$dlSetError'],
   _emscripten_dlopen_js: (handle, onsuccess, onerror, user_data) => {
-    /** @param {Object=} e */
+  return new Promise((resolve, reject) => {
     function errorCallback(e) {
-      var filename = UTF8ToString(handle + {{{ C_STRUCTS.dso.name }}});
-      dlSetError(`'Could not load dynamic lib: ${filename}\n${e}`);
-      {{{ runtimeKeepalivePop() }}}
-      callUserCallback(() => {{{ makeDynCall('vpp', 'onerror') }}}(handle, user_data));
+      var filename = UTF8ToString(handle + C_STRUCTS.dso.name);
+      dlSetError(`Could not load dynamic lib: ${filename}\n${e}`);
+      runtimeKeepalivePop();
+      if(onerror) {
+        callUserCallback(() => makeDynCall('vpp', 'onerror')(handle, user_data));
+      }
+      reject(e);
     }
     function successCallback() {
-      {{{ runtimeKeepalivePop() }}}
-      callUserCallback(() => {{{ makeDynCall('vpp', 'onsuccess') }}}(handle, user_data));
+      runtimeKeepalivePop();
+      if(onsuccess) {
+        callUserCallback(() => makeDynCall('vpp', 'onsuccess')(handle, user_data));
+      }
+      resolve();
     }
 
-    {{{ runtimeKeepalivePush() }}}
+    runtimeKeepalivePush();
     var promise = dlopenInternal(handle, { loadAsync: true });
```

Comment:
So this is an example of API that already has a promise being returned from the underlying API.

Do I think we have two reasonable ways forward to make this more generic:

1.  Return a `promise_t` back to C/C++ here and let C/C++ handler wrapping this into a callback-based API.
2. Create a generic JS wrapper function called something like `promiseToCallback(successFnPtr, errorFrPtr)` that takes care of doing all the `runtimeKeepalivePop` stuff and the `callUserCallback` stuff.

I think (2) might be the simplest for now as it probably involves a little less complexity since I believe the C/C++ promise API would end up creating an extra promise in order to do the `emscripten_promise_then` operation.

@tlively would be OK with (1) here?

Reply:
Did you mean (2)? Either sounds good to me 👍

Reply:
@sbc100 @tlively Made the changes according to the feedback provided above
1. Created a generic JS wrapper function called something like promiseToCallback(successFnPtr, errorFrPtr) that takes care of doing all the runtimeKeepalivePop stuff and the callUserCallback
2. Added {{{ and }}} around  calls.
3. Added asset(promise)  instead of if/elfse